### ISSUE TITLE
[BUG]  getting AssertionError from  ItemRetrievalScorer() class 

### ISSUE THREAD

**Issue by rnyak** [NVIDIA-Merlin/models#173]
I am getting the following error from training a mf-retrieval model. The issue is that ItemRetrieval scorer requires `query` and `item` tags from the schema, whereas they dont exist in the schema.pbtxt. This can be solved either
- user will define `user_id` as `query` and `item_id` as `item` features during the ETL process.
-  or needs to be fixed in the ItemRetrievalScorer(Block) class.

```
import merlin_models.tf as mm
mf = mm.MatrixFactorizationBlock(schema, dim=128)
model = mf.connect(mm.ItemRetrievalTask(samplers = [mm.InBatchSampler()]))
model.compile(optimizer="adam", run_eagerly=False)
train_loader = get_dataloader(nvt.Dataset(train_paths, part_size="256MB"), shuffle=True) 
losses = model.fit(train_loader, epochs=3)
model.reset_metrics()
```
```
2022-02-08 19:59:03.494259: W tensorflow/python/util/util.cc:368] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Input In [10], in <module>
     10 print('*'*20 + '\n')
     11 train_loader = get_dataloader(nvt.Dataset(train_paths, part_size="256MB"), shuffle=True) 
---> 12 losses = model.fit(train_loader, epochs=3)
     13 model.reset_metrics()

File /usr/local/lib/python3.8/dist-packages/keras/engine/training.py:1134, in Model.fit(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_batch_size, validation_freq, max_queue_size, workers, use_multiprocessing)
   1128   self._cluster_coordinator = tf.distribute.experimental.coordinator.ClusterCoordinator(
   1129       self.distribute_strategy)
   1131 with self.distribute_strategy.scope(), \
   1132      training_utils.RespectCompiledTrainableState(self):
   1133   # Creates a `tf.data.Dataset` and handles batch and epoch iteration.
-> 1134   data_handler = data_adapter.get_data_handler(
   1135       x=x,
   1136       y=y,
   1137       sample_weight=sample_weight,
   1138       batch_size=batch_size,
   1139       steps_per_epoch=steps_per_epoch,
   1140       initial_epoch=initial_epoch,
   1141       epochs=epochs,
   1142       shuffle=shuffle,
   1143       class_weight=class_weight,
   1144       max_queue_size=max_queue_size,
   1145       workers=workers,
   1146       use_multiprocessing=use_multiprocessing,
   1147       model=self,
   1148       steps_per_execution=self._steps_per_execution)
   1150   # Container that configures and calls `tf.keras.Callback`s.
   1151   if not isinstance(callbacks, callbacks_module.CallbackList):

File /usr/local/lib/python3.8/dist-packages/keras/engine/data_adapter.py:1383, in get_data_handler(*args, **kwargs)
   1381 if getattr(kwargs["model"], "_cluster_coordinator", None):
   1382   return _ClusterCoordinatorDataHandler(*args, **kwargs)
-> 1383 return DataHandler(*args, **kwargs)

File /usr/local/lib/python3.8/dist-packages/keras/engine/data_adapter.py:1138, in DataHandler.__init__(self, x, y, sample_weight, batch_size, steps_per_epoch, initial_epoch, epochs, shuffle, class_weight, max_queue_size, workers, use_multiprocessing, model, steps_per_execution, distribute)
   1135   self._steps_per_execution_value = steps_per_execution.numpy().item()
   1137 adapter_cls = select_data_adapter(x, y)
-> 1138 self._adapter = adapter_cls(
   1139     x,
   1140     y,
   1141     batch_size=batch_size,
   1142     steps=steps_per_epoch,
   1143     epochs=epochs - initial_epoch,
   1144     sample_weights=sample_weight,
   1145     shuffle=shuffle,
   1146     max_queue_size=max_queue_size,
   1147     workers=workers,
   1148     use_multiprocessing=use_multiprocessing,
   1149     distribution_strategy=tf.distribute.get_strategy(),
   1150     model=model)
   1152 strategy = tf.distribute.get_strategy()
   1154 self._current_step = 0

File /usr/local/lib/python3.8/dist-packages/keras/engine/data_adapter.py:917, in KerasSequenceAdapter.__init__(self, x, y, sample_weights, shuffle, workers, use_multiprocessing, max_queue_size, model, **kwargs)
    915 self._keras_sequence = x
    916 self._enqueuer = None
--> 917 super(KerasSequenceAdapter, self).__init__(
    918     x,
    919     shuffle=False,  # Shuffle is handed in the _make_callable override.
    920     workers=workers,
    921     use_multiprocessing=use_multiprocessing,
    922     max_queue_size=max_queue_size,
    923     model=model,
    924     **kwargs)

File /usr/local/lib/python3.8/dist-packages/keras/engine/data_adapter.py:801, in GeneratorDataAdapter.__init__(self, x, y, sample_weights, workers, use_multiprocessing, max_queue_size, model, **kwargs)
    799 if model is not None and not model.built:
    800   concrete_x, _, _ = unpack_x_y_sample_weight(peek)
--> 801   model.distribute_strategy.run(
    802       lambda x: model(x, training=False), args=(concrete_x,))
    804 self._first_batch_size = int(tf.nest.flatten(peek)[0].shape[0])
    806 def _get_dynamic_shape(t):

File /usr/local/lib/python3.8/dist-packages/tensorflow/python/distribute/distribute_lib.py:1316, in StrategyBase.run(***failed resolving arguments***)
   1311 with self.scope():
   1312   # tf.distribute supports Eager functions, so AutoGraph should not be
   1313   # applied when the caller is also in Eager mode.
   1314   fn = autograph.tf_convert(
   1315       fn, autograph_ctx.control_status_ctx(), convert_by_default=False)
-> 1316   return self._extended.call_for_each_replica(fn, args=args, kwargs=kwargs)

File /usr/local/lib/python3.8/dist-packages/tensorflow/python/distribute/distribute_lib.py:2892, in StrategyExtendedV1.call_for_each_replica(self, fn, args, kwargs)
   2890   kwargs = {}
   2891 with self._container_strategy().scope():
-> 2892   return self._call_for_each_replica(fn, args, kwargs)

File /usr/local/lib/python3.8/dist-packages/tensorflow/python/distribute/distribute_lib.py:3695, in _DefaultDistributionExtended._call_for_each_replica(self, fn, args, kwargs)
   3693 def _call_for_each_replica(self, fn, args, kwargs):
   3694   with ReplicaContext(self._container_strategy(), replica_id_in_sync_group=0):
-> 3695     return fn(*args, **kwargs)

File /usr/local/lib/python3.8/dist-packages/tensorflow/python/autograph/impl/api.py:601, in call_with_unspecified_conversion_status.<locals>.wrapper(*args, **kwargs)
    599 def wrapper(*args, **kwargs):
    600   with ag_ctx.ControlStatusCtx(status=ag_ctx.Status.UNSPECIFIED):
--> 601     return func(*args, **kwargs)

File /usr/local/lib/python3.8/dist-packages/keras/engine/data_adapter.py:802, in GeneratorDataAdapter.__init__.<locals>.<lambda>(x)
    799 if model is not None and not model.built:
    800   concrete_x, _, _ = unpack_x_y_sample_weight(peek)
    801   model.distribute_strategy.run(
--> 802       lambda x: model(x, training=False), args=(concrete_x,))
    804 self._first_batch_size = int(tf.nest.flatten(peek)[0].shape[0])
    806 def _get_dynamic_shape(t):

File /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:1037, in Layer.__call__(self, *args, **kwargs)
   1033   inputs = self._maybe_cast_inputs(inputs, input_list)
   1035 with autocast_variable.enable_auto_cast_variables(
   1036     self._compute_dtype_object):
-> 1037   outputs = call_fn(inputs, *args, **kwargs)
   1039 if self._activity_regularizer:
   1040   self._handle_activity_regularization(inputs, outputs)

File /models/merlin_models/tf/core.py:2043, in Model.call(self, inputs, **kwargs)
   2042 def call(self, inputs, **kwargs):
-> 2043     outputs = self.body(inputs, **kwargs)
   2044     return outputs

File /models/merlin_models/config/schema.py:50, in SchemaMixin.__call__(self, *args, **kwargs)
     47 def __call__(self, *args, **kwargs):
     48     self.check_schema()
---> 50     return super().__call__(*args, **kwargs)

File /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:1030, in Layer.__call__(self, *args, **kwargs)
   1028 with tf.name_scope(name_scope):
   1029   if not self.built:
-> 1030     self._maybe_build(inputs)
   1032   if self._autocast:
   1033     inputs = self._maybe_cast_inputs(inputs, input_list)

File /models/merlin_models/tf/core.py:203, in Block._maybe_build(self, inputs)
    200 if getattr(self, "_context", None) and not self.context.built:
    201     self.context.set_dtypes(inputs)
--> 203 super()._maybe_build(inputs)

File /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:2659, in Layer._maybe_build(self, inputs)
   2654 if not hasattr(self.build, '_is_default'):
   2655   # Any setup work performed only once should happen in an `init_scope`
   2656   # to avoid creating symbolic Tensors that will later pollute any eager
   2657   # operations.
   2658   with tf_utils.maybe_init_scope(self):
-> 2659     self.build(input_shapes)  # pylint:disable=not-callable
   2660 # We must set also ensure that the layer is marked as built, and the build
   2661 # shape is stored since user defined build functions may not be calling
   2662 # `super.build()`
   2663 Layer.build(self, input_shapes)

File /models/merlin_models/tf/core.py:476, in SequentialBlock.build(self, input_shape)
    471             v = TypeError(
    472                 f"Couldn't build {layer}, "
    473                 f"did you forget to add aggregation to {last_layer}?"
    474             )
    475         six.reraise(t, v, tb)
--> 476     input_shape = layer.compute_output_shape(input_shape)
    477     last_layer = layer
    478 self.built = True

File /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:783, in Layer.compute_output_shape(self, input_shape)
    781 inputs = tf.nest.map_structure(_make_placeholder_like, input_shape)
    782 try:
--> 783   outputs = self(inputs, training=False)
    784 except TypeError as e:
    785   raise NotImplementedError(
    786       'We could not automatically infer the static shape of the '
    787       'layer\'s output. Please implement the '
    788       '`compute_output_shape` method on your layer (%s).' %
    789       self.__class__.__name__) from e

File /models/merlin_models/tf/core.py:1611, in PredictionTask.__call__(self, *args, **kwargs)
   1610 def __call__(self, *args, **kwargs):
-> 1611     inputs = self.pre_call(*args, **kwargs)
   1613     # This will call the `call` method implemented by the super class.
   1614     outputs = super().__call__(inputs, **kwargs)  # noqa

File /models/merlin_models/tf/core.py:1602, in PredictionTask.pre_call(self, inputs, **kwargs)
   1599     x = self.task_block(x)
   1601 if self.pre:
-> 1602     x = self.pre(inputs, **kwargs)
   1604 return x

File /models/merlin_models/config/schema.py:50, in SchemaMixin.__call__(self, *args, **kwargs)
     47 def __call__(self, *args, **kwargs):
     48     self.check_schema()
---> 50     return super().__call__(*args, **kwargs)

File /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:976, in Layer.__call__(self, *args, **kwargs)
    970 # Functional Model construction mode is invoked when `Layer`s are called on
    971 # symbolic `KerasTensor`s, i.e.:
    972 # >> inputs = tf.keras.Input(10)
    973 # >> outputs = MyLayer()(inputs)  # Functional construction mode.
    974 # >> model = tf.keras.Model(inputs, outputs)
    975 if _in_functional_construction_mode(self, inputs, args, kwargs, input_list):
--> 976   return self._functional_construction_call(inputs, args, kwargs,
    977                                             input_list)
    979 # Maintains info about the `Layer.call` stack.
    980 call_context = base_layer_utils.call_context()

File /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:1114, in Layer._functional_construction_call(self, inputs, args, kwargs, input_list)
   1109     training_arg_passed_by_framework = True
   1111 with call_context.enter(
   1112     layer=self, inputs=inputs, build_graph=True, training=training_value):
   1113   # Check input assumptions set after layer building, e.g. input shape.
-> 1114   outputs = self._keras_tensor_symbolic_call(
   1115       inputs, input_masks, args, kwargs)
   1117   if outputs is None:
   1118     raise ValueError('A layer\'s `call` method should return a '
   1119                      'Tensor or a list of Tensors, not None '
   1120                      '(layer: ' + self.name + ').')

File /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:848, in Layer._keras_tensor_symbolic_call(self, inputs, input_masks, args, kwargs)
    846   return tf.nest.map_structure(keras_tensor.KerasTensor, output_signature)
    847 else:
--> 848   return self._infer_output_signature(inputs, args, kwargs, input_masks)

File /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:886, in Layer._infer_output_signature(self, inputs, args, kwargs, input_masks)
    880 with backend.name_scope(self._name_scope()):  # pylint: disable=not-callable
    881   with autocast_variable.enable_auto_cast_variables(
    882       self._compute_dtype_object):
    883     # Build layer if applicable (if the `build` method has been
    884     # overridden).
    885     # TODO(kaftan): do we maybe_build here, or have we already done it?
--> 886     self._maybe_build(inputs)
    887     inputs = self._maybe_cast_inputs(inputs)
    888     outputs = call_fn(inputs, *args, **kwargs)

File /models/merlin_models/tf/core.py:203, in Block._maybe_build(self, inputs)
    200 if getattr(self, "_context", None) and not self.context.built:
    201     self.context.set_dtypes(inputs)
--> 203 super()._maybe_build(inputs)

File /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:2659, in Layer._maybe_build(self, inputs)
   2654 if not hasattr(self.build, '_is_default'):
   2655   # Any setup work performed only once should happen in an `init_scope`
   2656   # to avoid creating symbolic Tensors that will later pollute any eager
   2657   # operations.
   2658   with tf_utils.maybe_init_scope(self):
-> 2659     self.build(input_shapes)  # pylint:disable=not-callable
   2660 # We must set also ensure that the layer is marked as built, and the build
   2661 # shape is stored since user defined build functions may not be calling
   2662 # `super.build()`
   2663 Layer.build(self, input_shapes)

File /models/merlin_models/tf/core.py:476, in SequentialBlock.build(self, input_shape)
    471             v = TypeError(
    472                 f"Couldn't build {layer}, "
    473                 f"did you forget to add aggregation to {last_layer}?"
    474             )
    475         six.reraise(t, v, tb)
--> 476     input_shape = layer.compute_output_shape(input_shape)
    477     last_layer = layer
    478 self.built = True

File /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:783, in Layer.compute_output_shape(self, input_shape)
    781 inputs = tf.nest.map_structure(_make_placeholder_like, input_shape)
    782 try:
--> 783   outputs = self(inputs, training=False)
    784 except TypeError as e:
    785   raise NotImplementedError(
    786       'We could not automatically infer the static shape of the '
    787       'layer\'s output. Please implement the '
    788       '`compute_output_shape` method on your layer (%s).' %
    789       self.__class__.__name__) from e

File /models/merlin_models/config/schema.py:50, in SchemaMixin.__call__(self, *args, **kwargs)
     47 def __call__(self, *args, **kwargs):
     48     self.check_schema()
---> 50     return super().__call__(*args, **kwargs)

File /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:976, in Layer.__call__(self, *args, **kwargs)
    970 # Functional Model construction mode is invoked when `Layer`s are called on
    971 # symbolic `KerasTensor`s, i.e.:
    972 # >> inputs = tf.keras.Input(10)
    973 # >> outputs = MyLayer()(inputs)  # Functional construction mode.
    974 # >> model = tf.keras.Model(inputs, outputs)
    975 if _in_functional_construction_mode(self, inputs, args, kwargs, input_list):
--> 976   return self._functional_construction_call(inputs, args, kwargs,
    977                                             input_list)
    979 # Maintains info about the `Layer.call` stack.
    980 call_context = base_layer_utils.call_context()

File /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:1114, in Layer._functional_construction_call(self, inputs, args, kwargs, input_list)
   1109     training_arg_passed_by_framework = True
   1111 with call_context.enter(
   1112     layer=self, inputs=inputs, build_graph=True, training=training_value):
   1113   # Check input assumptions set after layer building, e.g. input shape.
-> 1114   outputs = self._keras_tensor_symbolic_call(
   1115       inputs, input_masks, args, kwargs)
   1117   if outputs is None:
   1118     raise ValueError('A layer\'s `call` method should return a '
   1119                      'Tensor or a list of Tensors, not None '
   1120                      '(layer: ' + self.name + ').')

File /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:848, in Layer._keras_tensor_symbolic_call(self, inputs, input_masks, args, kwargs)
    846   return tf.nest.map_structure(keras_tensor.KerasTensor, output_signature)
    847 else:
--> 848   return self._infer_output_signature(inputs, args, kwargs, input_masks)

File /usr/local/lib/python3.8/dist-packages/keras/engine/base_layer.py:888, in Layer._infer_output_signature(self, inputs, args, kwargs, input_masks)
    886     self._maybe_build(inputs)
    887     inputs = self._maybe_cast_inputs(inputs)
--> 888     outputs = call_fn(inputs, *args, **kwargs)
    890   self._handle_activity_regularization(inputs, outputs)
    891 self._set_mask_metadata(inputs, outputs, input_masks,
    892                         build_graph=False)

File /usr/local/lib/python3.8/dist-packages/tensorflow/python/autograph/impl/api.py:699, in convert.<locals>.decorator.<locals>.wrapper(*args, **kwargs)
    697 except Exception as e:  # pylint:disable=broad-except
    698   if hasattr(e, 'ag_error_metadata'):
--> 699     raise e.ag_error_metadata.to_exception(e)
    700   else:
    701     raise

AssertionError: in user code:

    File "/models/merlin_models/tf/prediction/item_prediction.py", line 216, in call  *
        self._check_input_from_two_tower(inputs)
    File "/models/merlin_models/tf/prediction/item_prediction.py", line 171, in _check_input_from_two_tower  *
        assert set(inputs.keys()) == set(["query", "item"])

    AssertionError: 
```

**Comment by marcromeyn** [comment]
I would hope that we won’t need the user to specify names everywhere. Wouldn’t it be possible to add a post-process transform within the `MatrixFactorizationBlock` that would output the right keys?

**Comment by sararb** [comment]
Yeah, with Ronay and Gabriel, we discussed whether we fix the names of the branches within the `ItemRetrievalScorer` and make sure the user is setting the two-tower model blocks with the proper names (In that case, changing the names of the embeddings tensors given by the MatrixFactorization block). Or makes the `ItemRetrievalScorer` more generic and allows the user to change the names based on its use-case. 

In both cases, I think the user would need to specify the names if he defines a custom two-tower block. The latter solution could maybe offer more flexibility to the user? And at the same time, we could add the post-process op to the `MatrixFactorizationBlock` to facilitate the definition of the MF-retrieval model?
